Add Additional Features

tm1000 opened the pull request:
* Ability to list all Restaurants and list their associated ids
* Ability to send push notifications to pushover.net
* Ability to search for party size
* Ability to search multiple restaurants
* Add link to directly reserve from the notification

tm1000 committed:
Add Additional Features
-Ability to list all Restaurants and list their associated ids
-Ability to send push notifications to pushover.net
-Ability to search for party size
-Ability to search multiple restaurants
-Add link to directly reserve from the notification

tm1000 committed:
update readme

alechemy commented:
Wow, thanks for the PR! I am glad this project caught your interest, and maybe it was helpful to you. It is certainly half-baked as-is, so I appreciate the contributions.

I'm a bit busy right now but will try to review tomorrow, or Saturday at the latest.

tm1000 commented:
@alechemy Hey there! yeah actually I live very close to Disneyland and I found this project in my frustration to find reservations at the park after the covid lockdowns and additionally there was mention of another site in this article: https://www.ocregister.com/2021/10/18/nightmare-scenario-unfolds-as-disneyland-restaurant-reservations-snapped-up-months-in-advance/ which charges a fee (but also has a free tier) I figured if they were doing it as a paid service there must be something on github from someone... you have the only working and up to date project I was able to find. 

You can see in my code that I scrape the dining website to actually get the "PEP" values which then link to html elements which is how I can figure out the dining id and associate it to the real name of the restaurant. Anyways I've been using it for a couple of days and already got a few reservations.

I was also pretty impressed that you figured out how to get the Bearer token for authentication

jmiceter commented on src/extensions/cli-extension.ts:
```diff
@@ -21,7 +23,59 @@ const disneyApi: ApisauceInstance = http.create({
+  eval(finderBlob);
+
+  //Scan the DOM for the real name
+  return Object.values(PEP.Finder.List.cards).reduce((prev: any, card: any) => {
```
Suggestion. Instead of returning this object, I suggest filtering out only the reservation based restaurants. 
I tried this to no avial: await api.get('/dining/#/reservations-accepted/');

here is my method that finally worked for me:
```

  //Scan the DOM for the real name
  var locationsobj =  Object.values(PEP.Finder.List.cards).reduce((prev: any, card: any) => {
    const htmlCard = dom.window.document.querySelector(card.ref);
    return {
      ...prev,
      [card.id.split(';')[0]]: { ...card, displayName: htmlCard.querySelector('.cardName').innerHTML },
    };
  }, {});
  //Keep only the reservation locations
  const locationsArr = Object.entries(locationsobj);
  const locationsAdjusted = locationsArr.filter(([key, value]) => value['facets']['reservations-accepted'] !== undefined)
                                      .filter(([key, value]) => value['facets']['reservations-accepted']['value']%2 !== 0);
  return Object.fromEntries(locationsAdjusted);
```

tm1000 replied:
Looks good. I'll add this in the next week and additionally do a regular expression to pull out PEP instead of eval-ing it

jmiceter commented:
This is great. I was about to jump on these exact changes when I found this. I'm glad you beat me to it. I may add a park column and organize the listings in the future.

alechemy commented on README.md:
```diff
@@ -2,15 +2,84 @@
+
+#### Pushover alerts
+
+For more information see: https://pushover.net/
```
Could you make this a properly formatted Markdown link? e.g.
```suggestion
For more information see: [https://pushover.net/](https://pushover.net/)
```

L82 as well.

alechemy commented on src/extensions/cli-extension.ts:
```diff
@@ -30,60 +84,166 @@ async function getAuthToken(api: ApisauceInstance = disneyApi): Promise<any> {
-        // Seems like we found something
-        onSuccess(hasOffers);
+        if (ids) {
+          await eachSeries(ids, async (id: string, callback) => {
```
I am unfamiliar with the `async` package, but I wonder if we can achieve the desired behavior here without relying on a new package (esp. since you're just using the single operator).

I found this page on replacing the `async` lib with async/await, and it includes a specific example for `eachSeries`. 

https://github.com/jmjpro/async-package-to-async-await/blob/master/README.md

tm1000 replied:
Yeah you caught me being lazy (embarrassed). I really needed that reservation when I wrote this hahaha

alechemy commented on src/commands/search.ts:
```diff
@@ -1,19 +1,41 @@
+        date,
+        onSuccess,
+        print,
+        ids: (ids && typeof ids === 'number' ? ids.toString() : ids).split(','),
```
unnecessary to check `ids` existence again in this if block

```suggestion
        ids: (typeof ids === 'number' ? ids.toString() : ids).split(','),
```

tm1000 replied:
Oh right. I'm surprised typescript didn't complain at me about that. Although I'm sure eslint would

alechemy commented on src/extensions/cli-extension.ts:
```diff
@@ -21,7 +23,59 @@ const disneyApi: ApisauceInstance = http.create({
+  };
+};
+
+const memoziedRestaurantMapping = promiseMemoize(restaurantMapping);
```
typo here , I think

```suggestion
const memoizedRestaurantMapping = promiseMemoize(restaurantMapping);
```

alechemy reviewed (commented):
Thanks again for the contributions. I left a few comments, but LGTM overall. 😄

tm1000 committed:
fix outstanding issues

tm1000 commented:
@alechemy I've updated the PR with the requested changes from you and @jmiceter 

Additionally I dropped using eval and I'm just using JSON parse so the code is much safer now
